Dojo 6 Release Blog (#151)
dojo/site: content/blog/en/version-6-dojo.md

---
title: Announcing Dojo 6

> date: 2019-07-30T08:00:00.000Z

Reviewer: Does this need to be updated?

Author: Yup!

author: Anthony Gubler
---

Since the first major Dojo release last year, we have been working to refine the features and patterns to make Dojo an even more efficient framework for building applications with TypeScript and modern web APIs.

In a world of semantic versioning where even minor breaking changes require a new version number, it is challenging to know when a new version is substantial. Today, we’re excited to announce version 6, our most ambitious set of improvements since the Dojo 2.0 release.

> Version 6 of Dojo brings many new features and changes to substantially improve the development experience when creating applications with Dojo by reducing boilerplate, increasing flexibility, and further improving performance.

Reviewer: ```suggestion
Version 6 of Dojo brings many new features and changes to substantially improve the development experience by reducing boilerplate, increasing flexibility, and further improving performance.
```

![The image for the blog](assets/blog/version-6-dojo/featured.png)
<!-- more -->

> ## Function Based Widgets and Middleware

Reviewer: Not sure if we want to align on Sentence-style capitalization for headers, as per the ref guides

Author: I wasn't sure if blogs were different as I had a look at a couple of different release blogs and they seemed to have capitalised headings.

Author: Also all the previous blogs followed the same pattern

Reviewer: Yeah i suppose consistency with other blog entries trumps newer parts of the site.

Reviewer 2: We could always update the other blog entries.

> We are very excited to introduce function-based widgets and middleware, the next evolution for creating and working with widgets in Dojo. Function-based widgets and middleware offer an alternative API for Dojo widgets to the existing class-based APIs (metas, decorators, mixins). A single API for both widgets and middleware helps improve developer ergonomics and reduce the complexity and boilerplate with previous releases of Dojo, making it even easier to get started with Dojo.

Reviewer: ```suggestion
We are very excited to introduce function-based widgets and middleware, the next evolution for creating and working with widgets in Dojo. Function-based widgets and middleware offer an alternative API to the existing class-based APIs (metas, decorators, mixins). A single API for both widgets and middleware helps improve developer ergonomics and reduce the complexity and boilerplate compared to previous releases of the framework, making it even easier to start writing Dojo applications.
```

What is middleware? Middleware is the singular concept designed to replace all existing supplemental widget patterns, mixins, metas and decorators. A natural progression from the functional and reactive meta API, middleware provides a mechanism that not only facilitates working with imperative and asynchronous APIs reactively, but can get composed with other middleware and also affect a widget’s property interface.

The core primitive for working with function-based widgets and middleware is a new function, `create`, provided by the `vdom` module. The `create` function gets used for defining middleware and properties and returns a factory for creating either widgets or middleware.

> src/MyWidget.tsx
```tsx
import { create, tsx } from '@dojo/framework/core/vdom';

interface MyWidgetProperties {
	label: string;
}

> const render = create().properties<MyWidgetProperties>();

Reviewer 3: It might make sense to update this example to inject middleware as well since that is the primary thing that we're talking about above.

Author: The example below has a middleware injected, is that good enough? I was wary to try and crowbar middleware in without being really meaningful.

Reviewer 3: Yeah, that makes sense.

export default render(function MyWidget({ properties, children }) {
  const { label } = properties();
	return (
		<div>
			<span>{label}</span>
			<div>{children()}<div>
		</div>
	);
});
```

> As mentioned, the middleware design supports composition in order to create advanced custom functionality. The majority of middlewares will build on the set of core middlewares which provide hooks into Dojo’s rendering engine. For more information on the core middleware please see the [Dojo reference guide](/learn/middleware/core-render-middleware).

Reviewer 2: A small change to use active voice instead of passive here.

```suggestion
The middleware design supports composition in order to create advanced custom functionality. The majority of middlewares will build on the set of core middlewares which provide hooks into Dojo’s rendering engine. For more information on the core middleware please see the [Dojo reference guide](/learn/middleware/core-render-middleware).
```

> In addition to the core middleware, we've created a selection of [higher-level middleware](/learn/middleware/available-middleware) for function-based widgets. These middlewares provide the features and functionality found in most of the existing metas and mixins from working with class-based widgets.

Reviewer 4: ```suggestion
In addition to the core middleware, we've created a selection of [higher-level Dojo middleware](/learn/middleware/available-middleware) for function-based widgets. These middlewares provide the features and functionality found in most of the existing metas and mixins from working with class-based widgets.
```

> A common pattern used in class-based widgets is to use class properties to store local state for the widget. With function-based widgets this presents a challenge as there is no mechanism to persist state across widget renders. This is where the `icache` middleware, meaning invalidating cache, comes to the forefront. We believe `icache` being one of the most commonly middlewares when building function-based widgets.

Reviewer: ```suggestion
A common pattern used in class-based widgets is to use class properties to store local state for the widget. With function-based widgets this presents a challenge as there is no mechanism to persist state across widget renders. This is where the `icache` - invalidating cache - middleware comes to the forefront. We believe `icache` will be one of the most commonly used middleware when building function-based widgets.
```

        />
      </div>
    );
  }
}
```

Using function-based widgets, we define `icache` middleware for the widget factory so that we can store the counter state across renders. Notice that when we call set on the `icache` we don't need to manually `invalidate` the widget as `icache` implicitly invalidates when a value is set.

> src/FunctionalCounter.tsx
```tsx
interface FunctionalCounterProperties {
  incrementStep?: number;
}

const factory = create({ icache }).properties<FunctionalCounterProperties>();

export const FunctionalCounter = factory(function FunctionalCounter({
  middleware: { icache },
  properties
}) {
  const { incrementStep } = properties();
  const count = icache.get<number>("count") || 0;
  return (
    <div classes={[css.root]}>
      <button
        onclick={() => {
          icache.set("count", count - incrementStep);
        }}
        classes={[css.button, css.decrement]}
      />
      <div classes={[css.counter]}>{`${count}`}</div>
      <button
        onclick={() => {
          icache.set("count", count + incrementStep);
        }}
        classes={[css.button, css.increment]}
      />
    </div>
  );
});
```

> For more information please see the [middleware reference guides](/learn/middleware/introduction) or some of the middleware examples on codesandbox:

Reviewer 4: ```suggestion
For more information please see the [Dojo middleware reference guides](/learn/middleware/introduction) or some of the middleware examples on CodeSandbox:
```

>  * [`theme` middleware](https://codesandbox.io/s/theme-middleware-4btv7)

Reviewer 4: ```suggestion
 * [Dojo `theme` middleware](https://codesandbox.io/s/theme-middleware-4btv7)
```

>  * [`icache` middleware](https://codesandbox.io/s/advanced-icache-middleware-teeig)

Reviewer 4: ```suggestion
 * [Dojo `icache` middleware](https://codesandbox.io/s/advanced-icache-middleware-teeig)
```

Don't worry, the existing class-based widget APIs are not going away! These enhancements are additive and backwards-compatible, providing what we believe to be an ergonomic improvement on top of the existing widget APIs.

We are really looking forward to seeing the fun and innovative middlewares from the Dojo community, and, as always, please let us know any feedback that you might have!

## Config free Custom Elements

> Including Web Components as a first class citizen in Dojo is something that we’ve always been passionate about and now compiling your Dojo widgets to Custom Elements is _even_ easier. In Dojo 6 there is no configuration required, other than defining the widget(s) in the `.dojorc` to instruct the `@dojo/cli-build-widget` to compile your widgets to custom elements. We think this is a significant improvement to current tooling that requires widgets to be explicitly configured with the custom element details. Doing so takes additional development effort, foresight, and creates an additional maintenance burden of keeping the configuration up to date with changes to widget properties. This can be seen below with the custom element configuration for an example widget.

Reviewer 4: ```suggestion
Including Web Components as a first class citizen in Dojo is something that we’ve always been passionate about and now compiling your Dojo widgets to Custom Elements is _even_ easier. In Dojo 6 there is no configuration required, other than defining the widget(s) in the `.dojorc` to instruct the `@dojo/cli-build-widget` to compile your widgets to custom elements. We think this is a significant improvement to current tooling that requires widgets to be explicitly configured with the custom element details. The approach prior to Dojo version 6 required additional development effort, foresight, and created an additional maintenance burden of keeping the configuration up to date with changes to widget properties. This can be seen below with the custom element configuration for an example widget.
```

> Current configuration required for compiling a Dojo widget to a custom element:

Reviewer 4: ```suggestion
Configuration required prior to version 6 for compiling a Dojo widget to a custom element:
```

> .dojorc
```json
{
  "build-widget": {
    "widgets": [
      "src/MyWidget"
    ]
  }
}
```

> src/MyWidget
```tsx
import { WidgetBase } from '@dojo/framework/core/WidgetBase';
import { customElement } from '@dojo/framework/core/decorators/customElement';
import { CustomElementChildType } from '@dojo/framework/core/registerCustomElement';

interface MyWidgetProperties {
  value: string;
  disabled: boolean;
  onInput: (input: value) => void;
}

@customElement<MyWidgetProperties>({
	tag: 'dojo-my-widget',
	childType: CustomElementChildType.TEXT,
	properties: ['disabled'],
	attributes: ['value'],
	events: [
		'onInput'
	]
})
class MyWidget extends WidgetBase<MyWidgetProperties> {
  protected render() {
    // rendering
  }
}
```

> Using the build tool to intelligent configuration, requires only the `.dojorc` entry and will automatically include new, changed or removed properties.

Reviewer: ```suggestion
In Dojo version 6, the build tool's intelligent configuration of custom elements only requires the `.dojorc` entry and will automatically include new, changed or removed properties.
```

Might also be worth including a function-based equivalent of `MyWidget` to highlight reduced code config (i.e. no `@customElement`)?

Author: I don't think so, because it could be a class-based or function-based widget. So seeing the previous configuration and reading that it is no longer required should be enough?

> In Dojo version 6, the build tool's intelligent configuration of custom elements only requires the `.dojorc` entry and will automatically include new, changed or removed properties.

Reviewer 4: ```suggestion
In Dojo version 6, this is substantially simplified. Now the Dojo build tool's intelligent configuration of custom elements only requires the `.dojorc` entry and will automatically include new, changed or removed properties!
```

> In Dojo version 6, this is substantially simplified. Now the Dojo build tool's intelligent configuration of custom elements only requires the `.dojorc` entry and will automatically include new, changed or removed properties!

Reviewer 3: I love this old v new pattern that is used throughout this post. It not only highlights what changed but gives a clear direction on how to adopt those changes.

> .dojorc
```json
{
  "build-widget": {
    "widgets": [
      "src/MyWidget"
    ]
  }
}
```

## BTR and Dojo Block Improvements

In version 5 of Dojo, we announced Dojo Blocks, a feature leveraging build time rendering (BTR) that brought the world of static site generation to Dojo. Since then we have been working on improving the experience including more intelligent block bundling, dynamic path registration and a full static mode. Building static and progressive websites has never been easier with Dojo.

> Please check out our [example static blog site with Dojo on codesandbox](https://codesandbox.io/s/my-first-blog-bywnn).

Reviewer 4: ```suggestion
Review the [example static blog site with Dojo on CodeSandbox](https://codesandbox.io/s/my-first-blog-bywnn) to learn more about BTR and Dojo Block improvements.
```

> 

Reviewer 5: With Dojo 6, BTR has gotten even smarter. During the build, BTR will figure out what scripts and CSS assets were used and add them to the rendered output as preload tags. Downloading these important files right away, and before they are needed by the page, results in faster page performance. Additionally, assets that are not used by the current page, but are used by other pages, are added to the page as "prefetch", which instructs the browser to download these files when it's idle. Once the user interacts and chooses one of these other pages, the assets for the page will have already been loaded by the browser's prefetch. This results in a substantial load-time improvement.

Author: @rorticus Added. Thanks!

## Widget Library Build

A long awaited and highly requested feature for Dojo has been support for building Dojo widget libraries using the `@dojo/cli-build-widget` command. As part of Dojo 6 we're excited to include a library target for the first time. We are now using this to build the `@dojo/widgets` library and are very excited to see more Dojo widget libraries popping up throughout the community ❤️.

To build your widgets using `@dojo/cli-build-widget`, list your widgets in the `widgets` section of the `.dojorc` and run the build using the `--lib` option. The resulting build output is ready to be packaged and consumed by other Dojo applications.

> .dojorc
```json
{
  "build-widget": {
    "widgets": [
      "src/MyWidget"
    ]
  }
}
```

> terminal
```shell
dojo build widget --lib
```

> ## Faster Development Builds

Reviewer 3: Can we provide any info on how this is done? Are we intelligently detecting which files to recompile? Maybe it's too complex to mention here, but a bit of detail might be helpful.

Author: We don't tend to go into much detail around the implementation as these things can change, so probably best to keep it light on that front.

> As projects grow in size their build times can significantly increase. The CLI build command now supports an experimental “fast” mode that can reduce the build time of larger projects while developing.

Reviewer 4: ```suggestion
As projects grow in size their build times can significantly increase. The CLI build command now supports an experimental “fast” mode that can reduce the build time of larger projects during development.
```

> ![experimental speed demo](assets/blog/version-6-dojo/speed.gif)

Reviewer 4: ```suggestion
![experimental Dojo speed demo](assets/blog/version-6-dojo/speed.gif)
```

> ![experimental speed demo](assets/blog/version-6-dojo/speed.gif)

Author: I can do that...

> ![experimental speed demo](assets/blog/version-6-dojo/speed.gif)

Author: @sbinge done

> This shows an approximate saving of over 2 seconds building the [Dojo RealWorld example](https://github.com/dojo/examples/tree/master/realworld) with the experimental speed mode enabled. On larger Dojo projects we've witnessed more significant savings with development build times dropping by more than 50% having previously taking over 40 seconds, reducing to under 20 seconds using the new speed mode.

Reviewer 4: ```suggestion
This demonstrates saving approximately two seconds when building the [Dojo RealWorld example](https://github.com/dojo/examples/tree/master/realworld) with the experimental speed mode enabled. On larger Dojo projects we've witnessed more significant savings with development build times dropping by more than 50%, e.g. previous builds of more than 40 seconds reduced to under 20 seconds using the new speed mode.
```

## Glob Support For Code Splitting

> The `.dojorc` configuration for `bundles` has been enhanced to support globs. This is especially useful for scenarios such as internationalization, meaning that you don't have to define all the language bundles explicitly, instead simply define a matching pattern for each of the locales.

Reviewer 4: ```suggestion
The `.dojorc` configuration for `bundles` has been enhanced to support globs. The globs configuration option is especially useful for scenarios such as internationalization, meaning that you do not need to define all language bundles explicitly, instead simply define a matching pattern for each of the locales.
```

> Consider a project with a folder structure that defines language bundles in locale named directories, ideally build tool would create a single bundle for each locale that would be loaded when users change locale. Using a glob provides a low maintenance and minimal effort way to do this over explicitly defining each language bundle modules in the `.dojorc`.

Reviewer 4: ```suggestion
Consider a project with a folder structure that defines language bundles in locale named directories. Ideally the build tool would create a single bundle for each locale that gets loaded when the user changes their locale. Using a glob provides a low maintenance and minimal effort way to do this without explicitly defining every language bundle module in the `.dojorc`.
```

> Current configuration required to create a bundle per locale:

Reviewer 4: ```suggestion
Prior to Dojo version 6, the configuration required to create a bundle per locale:
```

> .dojorc
```json
{
  "build-app": {
    "bundles": {
      "fr": [ 
        "src/widgets/home/nls/fr/home",
        "src/widgets/menu/nls/fr/menu",
        "src/widgets/blog/nls/fr/blog",
        "src/widgets/reference/nls/fr/reference"
      ],
      "jp": [ 
        "src/widgets/home/nls/jp/home",
        "src/widgets/menu/nls/jp/menu",
        "src/widgets/blog/nls/jp/blog",
        "src/widgets/reference/nls/jp/reference"
      ]
    }
  }
}
```

> Using the glob configuration to create a bundle per locale:

Reviewer 4: ```suggestion
Using the glob configuration in Dojo version 6 to create a bundle per locale:
```

> .dojorc
```json
{
  "build-app": {
    "bundles": {
      "fr": [ "**/fr/*" ],
      "jp": [ "**/jp/*" ]
    }
  }
}
```

## Update to TypeScript

> The minimum required TypeScript version has been updated to 3.4. Updating the core framework to use a recent version enables us to leverage the latest features that underpin the function-based widget and middleware typings.

Reviewer 4: Should we mention that this is the first time we've changed this in over a year and give some guidance that we similarly don't plan to increment the minimum version for about the same amount of time unless absolutely necessary?

Author: I don't think so personally.

Reviewer 4: ok

Reviewer 2: dojo/framework's readme says the minimum is `3.4.5` versus this implies you could use `3.4.0`.

## Revamped Doc Website

// TODO talk about the creation of the new doc website - using BTR and Blocks

## Migration

> As usual all of the breaking changes introduced in Dojo 6 are carefully considered, so that we truly believe the benefits outweigh the upgrade effort. To assist with the transition we have updated the CLI upgrade command, which will automatically upgrade your Dojo dependencies, upgrade your application code where possible and highlight areas in the application that require manual intervention. For more information on what has changed in Dojo 6, please see the [migration guide](https://github.com/dojo/framework/blob/master/docs/V6-Migration-Guide.md).

Reviewer 4: ```suggestion
As usual all of the breaking changes introduced in Dojo 6 are carefully considered, so that we truly believe the benefits outweigh the upgrade effort. To assist with the transition we have updated the CLI upgrade command, which will automatically upgrade your Dojo dependencies, upgrade your application code where possible and highlight areas in the application that require manual intervention. For more information on what has changed in Dojo 6, please see the [Dojo version 6 migration guide](https://github.com/dojo/framework/blob/master/docs/V6-Migration-Guide.md).
```

Author: @dylans Do we need to repeat dojo 6 three words after saying dojo 6?

Reviewer 4: My main reason for making sure we include Dojo or Dojo version 6 in the links is to help search engines with the relevancy of the links, and to help with link scannability. We could of course remove it from earlier instances, though it probably does help with anyone trying to translate this content also.

## Support

> See the [release notes](TODO: Add link) for more details on version 6.0.0 of Dojo!

Reviewer 4: ```suggestion
See the [Dojo version 6 release notes](TODO: Add link) for more details on version 6.0.0 of Dojo!
```

> Love what we’re doing or having a problem? We ❤️our community. [Reach out to us on Discord](https://discord.gg/M7yRngE), check out [the Dojo roadmap](/roadmap) and see where we are heading, and try out the new [Dojo on CodeSandbox][https://codesandbox.io/s/github/dojo/dojo-codesandbox-template]. We look forward to your feedback!

Reviewer 4: ```suggestion
Love what we’re doing or having a problem? We ❤️our community. [Reach out to the Dojo team on Discord](https://discord.gg/M7yRngE), check out [the Dojo roadmap](/roadmap) and see where we are heading, and try out the new [Dojo on CodeSandbox][https://codesandbox.io/s/github/dojo/dojo-codesandbox-template]. We look forward to your feedback!
```

Prior to Dojo version 6, the configuration required to create a bundle per locale:

> .dojorc
```json
{
  "build-app": {
    "bundles": {
      "fr": [ 
        "src/widgets/home/nls/fr/home",
        "src/widgets/menu/nls/fr/menu",
        "src/widgets/blog/nls/fr/blog",
        "src/widgets/reference/nls/fr/reference"
      ],
      "jp": [ 
        "src/widgets/home/nls/jp/home",
        "src/widgets/menu/nls/jp/menu",
        "src/widgets/blog/nls/jp/blog",
        "src/widgets/reference/nls/jp/reference"
      ]
    }
  }
}
```

Using the glob configuration in Dojo version 6 to create a bundle per locale:

> .dojorc
```json
{
  "build-app": {
    "bundles": {
      "fr": [ "**/fr/*" ],
      "jp": [ "**/jp/*" ]
    }
  }
}
```

## Update to TypeScript

The minimum required TypeScript version has been updated to 3.4. Updating the core framework to use a recent version enables us to leverage the latest features that underpin the function-based widget and middleware typings.

## Revamped Doc Website

> For the Dojo 6 release, we have revamped [dojo.io](https://dojo.io) and moved it to a new repository, [dojo/site](https://github.com/dojo/site). Dojo.io now leverages the improved build time rendering (BTR) and block middleware. Documentation content is pulled down from [dojo/framework](https://github.com/dojo/framework) during the regular Dojo build process and rendered statically. The site is written completely using the new function-based widgets and middlewares, serving as a real world example of a statically rendered site using the latest Dojo patterns.

Reviewer 4: ```suggestion
For the Dojo 6 release, we have revamped [dojo.io](https://dojo.io) and moved it to a new repository, [dojo/site](https://github.com/dojo/site). Dojo.io now leverages the improved build time rendering (BTR) and block middleware. Documentation content is automatically imported from [dojo/framework](https://github.com/dojo/framework) during the regular Dojo build process and rendered statically. The new Dojo website is authored using the new Dojo function-based widgets and middlewares, serving as a real world example of a statically rendered site using the latest Dojo patterns!
```

## Migration

> As usual all of the breaking changes introduced in Dojo 6 are carefully considered, so that we truly believe the benefits outweigh the upgrade effort. To assist with the transition we have updated the CLI upgrade command, which will automatically upgrade your Dojo dependencies, upgrade your application code where possible and highlight areas in the application that require manual intervention. For more information on what has changed in Dojo 6, please see the [Dojo version 6 migration guide](https://github.com/dojo/framework/blob/master/docs/V6-Migration-Guide.md).

Reviewer 4: ```suggestion
As usual with modern Dojo releases,  all breaking changes introduced in Dojo 6 are carefully considered, so that we truly believe the benefits outweigh the upgrade effort. To assist with the transition we have updated the CLI upgrade command, which will automatically upgrade your Dojo dependencies, upgrade your application code where possible and highlight areas in the application that require manual intervention. For more information on what has changed in Dojo 6, please see the [Dojo version 6 migration guide](https://github.com/dojo/framework/blob/master/docs/V6-Migration-Guide.md).
```

Author: @dylans can you change the suggestion to remove the double space after `releases,`?
